source/common/config/type_url_loader.h

```diff
@@ -0,0 +1,88 @@
+#pragma once
+
+#include <string>
+#include <type_traits>
+
+#include "envoy/api/v2/auth/cert.pb.h"
+#include "envoy/api/v2/cluster.pb.h"
+#include "envoy/api/v2/endpoint.pb.h"
+#include "envoy/api/v2/listener.pb.h"
+#include "envoy/api/v2/rds.pb.h"
+#include "envoy/api/v2/route/route_components.pb.h"
+#include "envoy/api/v2/scoped_route.pb.h"
+#include "envoy/config/cluster/v3alpha/cluster.pb.h"
+#include "envoy/config/discovery_service_base.h"
+#include "envoy/config/endpoint/v3alpha/endpoint.pb.h"
+#include "envoy/config/listener/v3alpha/listener.pb.h"
+#include "envoy/config/route/v3alpha/route.pb.h"
+#include "envoy/config/route/v3alpha/route_components.pb.h"
+#include "envoy/config/route/v3alpha/scoped_route.pb.h"
+#include "envoy/extensions/transport_sockets/tls/v3alpha/cert.pb.h"
+#include "envoy/service/discovery/v2/rtds.pb.h"
+#include "envoy/service/runtime/v3alpha/rtds.pb.h"
+
+namespace Envoy {
+namespace Config {
+
+template <typename DiscoveryServiceSubscription> std::string v2ResourceNameEmitter() {
+  if (std::is_base_of<RdsRouteConfigSubscriptionBase, DiscoveryServiceSubscription>::value) {
+    return envoy::api::v2::RouteConfiguration().GetDescriptor()->full_name();
+  } else if (std::is_base_of<ScopedRdsConfigSubscriptionBase,
```

Comment:
This is a nice cleanup! I'm curious why we use this template driven approach, when we could just have all the subscriptions implement a v2/v3 descriptor name method. That would localize this information to the subscriptions.
/wait

Reply:
@htuch I think that this information should not be localized and should be used globally. And, it may eliminate `resources.h` completely. As a result, when we did dynamic version rewriting e.g. fallback, can be completed only to go through `type_url_loader.h`.
About template driven approach, there is enum based approach, string based approach and adopted approach. Simultaneously, I think that it is pretty good if there is same interface class about xDS that have understandable name e.g. `RdsRouteConfigSubscriptionBase`. 
I applied this approach because of these combination.

Reply:
Why can't `ScopedRdsConfigSubscriptionBase` just have a method `getResourceName()` that provides this information?